docker/webapp/app.py

```diff
@@ -0,0 +1,258 @@
+        self.download_url = f"/packages/{name}-{self.version}." + (entry["type"] == 'single' and "el" or "tar")
+        props = entry.get("props", {})
+        self.commit = props.get("commit")
+        self.dependencies = [Dependency(name, version_string(ver)) for (name, ver) in (entry["deps"] or {}).items()]
+        self._search_extra = recipe.get("repo")
+        self.source_url = calculate_source_url(recipe, self.commit)
+        self.home_url = props.get("url", self.source_url)
+        self.readme_url = f"/packages/{name}-readme.txt"
+        self.badge_url = f"/packages/{name}-badge.svg"
+        self.log_url = f"/packages/{name}.log"
+
+    def search_text(self):
+        return " ".join(x for x in [self.name, self.description, self.version, self._search_extra] if x)
+
+def maybe_timestamp(t):
+    if t is not None: return datetime.datetime.fromtimestamp(t)
+
+def maybe_duration(t):
+    if t is not None: return datetime.timedelta(seconds=t)
+
+class PackageData:
+    def __init__(self):
+        # Load raw data
+        with open(os.path.join(HTML_DIR, "recipes.json"), 'r') as f:
+            _recipes = json.load(f)
+        with open(os.path.join(HTML_DIR, "archive.json"), 'r') as f:
+            _archive = json.load(f)
+        with open(os.path.join(HTML_DIR, "download_counts.json"), 'r') as f:
+            _download_counts = json.load(f)
+        with open(os.path.join(HTML_DIR, "build-status.json"), 'r') as f:
+            _build_status = json.load(f)
+
+        # Assemble package info
+        self.packages_by_name = {name: PackageDescriptor(name, entry=entry,
+                              recipe=_recipes[name],
+                              download_counts=_download_counts)
+            for (name, entry) in _archive.items()}
+
+        self.packages = sorted(self.packages_by_name.values(),
+                               key=lambda p: p.name) # Pre-sort by the default case
+
+        # Pre-calculate download percentiles
+        by_downloads = sorted(self.packages, key=lambda p: p.downloads)
+        percentile = 0
+        self.download_percentiles = {}
+        for (_, cohort) in groupby(by_downloads, lambda p: p.downloads):
+            cohort = list(cohort)
+            self.download_percentiles.update({ p.name: percentile / len(by_downloads) for p in cohort })
+            percentile += len(cohort) * 100
+
+        self.total_downloads = sum(_download_counts.values())
+        self.last_build = BuildStatus(started_at=maybe_timestamp(_build_status["started"]),
+                                      completed_at=maybe_timestamp(_build_status["completed"]),
+                                      next_at=maybe_timestamp(_build_status["next"]),
+                                      duration=maybe_duration(_build_status["duration"]))
+    def is_stale(self):
+        return False
+
+package_data = None
+def load_package_data():
+    global package_data
```

Comment:
Consider using [`@lru_cache()`](https://docs.python.org/3/library/functools.html#functools.lru_cache) instead of a `global` variable. That way the cache will be stored in the function's object instead of in the `app.py`'s namespace and there would be no need for the conditions. Cleaning then is as simple as `load_package_data.cache_clear()`.

Reply:
Yay, thanks, this is a great tip. I've been mostly out of the Python ecosystem for a long time since the early days when I wrote `unittest.py`.

Reply:
I try mentioning it whenever I see singleton-like approach in Python which isn't heavy for performance. Less testing and bugs. If it were a possible bottleneck, maybe try it with timeit beforehands, though I believe the difference with the recent Python might be optimized away anyway. :blush: